File: src/models/Cliente.java
package models;

public class Cliente {
    public enum TipoCliente {
        NORMAL,
        DESCONHECIDO,
        SEMANAL,
    }

    private String nome;
    private String numeroTelefone;
    private TipoCliente tipoCliente;
    private int faltas;
    private String diaSemana;
    private String horaCorte;
    private boolean rapido;

    // Construtores
    public Cliente() {
        this.nome = "";
        this.numeroTelefone = "";
        this.tipoCliente = TipoCliente.DESCONHECIDO;
        this.faltas = 0;
        this.diaSemana = null;
        this.horaCorte = null;
        this.rapido = false;
    }

    // Construtor para NORMAL e DESCONHECIDO
    public Cliente(String nome, String numeroTelefone, TipoCliente tipoCliente) {
        this.nome = nome;
        this.numeroTelefone = numeroTelefone;
        this.tipoCliente = tipoCliente;
        this.faltas = 0;
        this.diaSemana = null;
        this.horaCorte = null;
        this.rapido = false;
    }

    // Construtor para SEMANAL
    public Cliente(String nome, String numeroTelefone, TipoCliente tipoCliente, String diaSemana, String horaCorte, boolean rapido) {
        this.nome = nome;
        this.numeroTelefone = numeroTelefone;
        this.tipoCliente = tipoCliente;
        this.faltas = 0;
        this.diaSemana = diaSemana;
        this.horaCorte = horaCorte;
        this.rapido = rapido;
    }

    public Cliente(Cliente outro) {
        this.nome = outro.nome;
        this.numeroTelefone = outro.numeroTelefone;
        this.tipoCliente = outro.tipoCliente;
        this.faltas = outro.faltas;
        this.diaSemana = outro.diaSemana;
        this.horaCorte = outro.horaCorte;
        this.rapido = outro.rapido;
    }

    // Getters e Setters
    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public String getNumeroTelefone() {
        return numeroTelefone;
    }

    public void setNumeroTelefone(String numeroTelefone) {
        this.numeroTelefone = numeroTelefone;
    }

    public TipoCliente getTipoCliente() {
        return tipoCliente;
    }

    public void setTipoCliente(TipoCliente tipoCliente) {
        this.tipoCliente = tipoCliente;
    }

    public int getFaltas() {
        return faltas;
    }

    public void setFaltas(int faltas) {
        this.faltas = faltas;
    }

    public String getDiaSemana() {
        return diaSemana;
    }

    public void setDiaSemana(String diaSemana) {
        this.diaSemana = diaSemana;
    }

    public String getHoraCorte() {
        return horaCorte;
    }

    public void setHoraCorte(String horaCorte) {
        this.horaCorte = horaCorte;
    }

    public boolean isRapido() {
        return rapido;
    }

    public void setRapido(boolean rapido) {
        this.rapido = rapido;
    }

    public boolean isTemporario() {
        return this.tipoCliente == TipoCliente.DESCONHECIDO;
    }

    @Override
    public Cliente clone() {
        return new Cliente(this);
    }

    @Override
    public String toString() {
        return "Cliente{" +
                "nome='" + nome + '\'' +
                ", numeroTelefone='" + numeroTelefone + '\'' +
                ", tipoCliente='" + tipoCliente + '\'' +
                ", faltas=" + faltas +
                (tipoCliente == TipoCliente.SEMANAL
                 ? ", diaSemana='" + diaSemana + '\'' +
                  ", horaCorte='" + horaCorte + '\''
                  : "") +
                ", rapido=" + rapido +
                '}';
    }
}

File: backend/models/Cliente.py
from enum import Enum

class TipoCliente(Enum):
    NORMAL = "NORMAL"
    DESCONHECIDO = "DESCONHECIDO"
    SEMANAL = "SEMANAL"
    
class Cliente:
    def __init__(self, nome="", numero_telefone="", tipo_cliente=TipoCliente.DESCONHECIDO,
                 dia_semana=None, hora_corte=None, rapido=False, outro=None):
        """
        Construtor da classe Cliente
        Pode ser usado como:
        - Cliente() - construtor padrão
        - Cliente(nome, numero_telefone, tipo_cliente) - para NORMAL e DESCONHECIDO
        - Cliente(nome, numero_telefone, TipoCliente.SEMANAL, dia_semana, hora_corte, rapido) - para SEMANAL
        - Cliente(outro=outro_cliente) - construtor de cópia
        """
        if outro is not None:
            # Construtor de cópia
            self.nome = outro.nome
            self.numero_telefone = outro.numero_telefone
            self.tipo_cliente = outro.tipo_cliente
            self.faltas = outro.faltas
            self.dia_semana = outro.dia_semana
            self.hora_corte = outro.hora_corte
            self.rapido = outro.rapido
        else:
            # Construtor padrão ou com parâmetros
            self.nome = nome
            self.numero_telefone = numero_telefone
            self.tipo_cliente = tipo_cliente
            self.faltas = 0
            self.dia_semana = dia_semana
            self.hora_corte = hora_corte
            self.rapido = rapido
            
    # Getters e Setters
    def get_nome(self):
        return self.nome
    
    def set_nome(self, nome):
        self.nome = nome
        
    def get_numero_telefone(self):
        return self.numero_telefone
    
    def set_numero_telefone(self, numero_telefone):
        self.numero_telefone = numero_telefone
        
    def get_tipo_cliente(self):
        return self.tipo_cliente
    
    def set_tipo_cliente(self, tipo_cliente):
        self.tipo_cliente = tipo_cliente
        
    def get_faltas(self):
        return self.faltas
    
    def set_faltas(self, faltas):
        self.faltas = faltas
        
    def get_dia_semana(self):
        return self.dia_semana
    
    def set_dia_semana(self, dia_semana):
        self.dia_semana = dia_semana
        
    def get_hora_corte(self):
        return self.hora_corte
    
    def set_hora_corte(self, hora_corte):
        self.hora_corte = hora_corte
        
    def is_rapido(self):
        return self.rapido
    
    def set_rapido(self, rapido):
        self.rapido = rapido
        
    def is_temporario(self):
        return self.tipo_cliente == TipoCliente.DESCONHECIDO
    
    def clone(self):
        return Cliente(outro=self)
    
    